test/hotspot/jtreg/compiler/lib/verify/Verify.java

```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+            print(a, b, field, aParent, bParent);
             throw new VerifyException("Value mismatch: " + (int)a + " vs " + (int)b);
         }
     }
 
     /**
      * Verify that two ints are identical.
      */
-    private static void checkEQimpl(int a, int b, String context) {
+    private void checkEQimpl(int a, int b, String field, Object aParent, Object bParent) {
         if (a != b) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + a + " vs " + b + " for " + context);
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + a + " vs " + b);
+            print(a, b, field, aParent, bParent);
             throw new VerifyException("Value mismatch: " + a + " vs " + b);
         }
     }
 
     /**
      * Verify that two longs are identical.
      */
-    private static void checkEQimpl(long a, long b, String context) {
+    private void checkEQimpl(long a, long b, String field, Object aParent, Object bParent) {
         if (a != b) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + a + " vs " + b + " for " + context);
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + a + " vs " + b);
+            print(a, b, field, aParent, bParent);
             throw new VerifyException("Value mismatch: " + a + " vs " + b);
         }
     }
 
     /**
-     * Verify that two floats have identical bits.
+     * There are two comparison modes: one where we compare the raw bits, which sees different NaN
+     * encodings as different values, and one where we see all NaN encodings as identical.
+     * Ideally, we would want to assert that the Float.floatToRawIntBits are identical.
+     * But the Java spec allows us to return different bits for a NaN, which allows swapping the inputs
+     * of an add or mul (NaN1 * NaN2 does not have same bits as NaN2 * NaN1, because the multiplication
+     * of two NaN values should always return the first of the two).
+     * Hence, by default, we pick the non-raw coparison: we verify that we have the same bit
+     * pattern in all cases, except for NaN we project to the canonical NaN, using Float.floatToIntBits.
      */
-    private static void checkEQimpl(float a, float b, String context) {
-        if (Float.floatToRawIntBits(a) != Float.floatToRawIntBits(b)) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch for " + context);
-            System.err.println("       Values: " + a + " vs " + b);
-            System.err.println("       Values: " + Float.floatToRawIntBits(a) + " vs " + Float.floatToRawIntBits(b));
+    private boolean isFloatEQ(float a, float b) {
+        return isFloatCheckWithRawBits ? Float.floatToRawIntBits(a) != Float.floatToRawIntBits(b)
+                                       : Float.floatToIntBits(a) != Float.floatToIntBits(b);
+    }
+
+    /**
+     * See comments for "isFloatEQ".
```

Comment:
We don't have Javadocs for the private methods but it could still help when navigating in the IDE to directly jump the the method when clicking on it (suggested same below for other places):
```suggestion
     * See comments for {@link #isFloatEQ}.
```